### PR TITLE
REVSDL-2456 OnServiceDeactivate notification GPB

### PR DESCRIPTION
**Pull request by giangtqh**
@bonewell 
@KKarlash 
@thinhts 
@DKlimenko 
@VladSemenyuk 
please review

### CODE REVIEW

**Comment by bonewell** [comment]
Reviewed.

File: src/components/vr_cooperation/src/on_service_deactivated_notification.cc [hunk 1]
```diff
@@ -40,14 +44,31 @@ namespace commands {
...
+      message_(message) {
 }
 
-void OnServiceDeactivatedNotification::Execute(
-    const application_manager::MessagePtr& message) {
+void OnServiceDeactivatedNotification::Run() {
   LOG4CXX_AUTO_TRACE(logger_);
-  parent_->SendMessageToMobile(message);
+  application_manager::MessagePtr mobile_msg = new application_manager::Message(
+        protocol_handler::MessagePriority::kDefault);
```
Comment (bonewell) [inline src/components/vr_cooperation/src/on_service_deactivated_notification.cc]: Wrong indent.

**Comment by DKlimenko** [comment]
reviewed

File: src/components/vr_cooperation/src/hmi_command_factory.cc [hunk 2]
```diff
@@ -46,10 +46,9 @@ commands::Command* HMICommandFactory::CreateCommand(
     const vr_hmi_api::ServiceMessage& message) {
   LOG4CXX_AUTO_TRACE(logger_);
   switch (message.rpc()) {
-// TODO(Giang): Uncomment when OnServiceDeactivatedNotification implementation
-//    case vr_hmi_api::RPCName::ON_DEACTIVATED: {
-//      return new commands::OnServiceDeactivatedNotification(parent, message);
-//    }
+    case vr_hmi_api::RPCName::ON_DEACTIVATED: {
+      return new commands::OnServiceDeactivatedNotification(parent, message);
```
Comment (KKarlash) [inline src/components/vr_cooperation/src/hmi_command_factory.cc]: could be memory leak

**Comment by KKarlash** [comment]
reviewed

File: src/components/vr_cooperation/include/vr_cooperation/commands/on_service_deactivated_notification.h [hunk 3]
```diff
@@ -35,27 +35,35 @@
...
-   * @param message message to be sent to mobile
+   * @brief run command
    */
-  void Execute(const application_manager::MessagePtr& message);
+  virtual void Run();
+
+  /**
+   * @brief on timeout reaction
+   */
+  virtual void OnTimeout();
```
Comment (KKarlash) [inline src/components/vr_cooperation/include/vr_cooperation/commands/on_service_deactivated_notification.h]: OnTimeout is not necessary for notification, that's why please make another base class special for notifications

File: src/components/vr_cooperation/src/hmi_command_factory.cc (hunk 2 above)
Comment (giangtqh) [inline src/components/vr_cooperation/src/hmi_command_factory.cc]: will be delete in VRModule

**Comment by bonewell** [comment]
Reviewed.

File: src/components/vr_cooperation/src/on_service_deactivated_notification.cc [hunk 4]
```diff
@@ -40,14 +44,31 @@ namespace commands {
...
   LOG4CXX_AUTO_TRACE(logger_);
-  parent_->SendMessageToMobile(message);
+  application_manager::MessagePtr mobile_msg = new application_manager::Message(
+      protocol_handler::MessagePriority::kDefault);
+  mobile_msg->set_function_id(
+      functional_modules::MobileFunctionID::ON_SERVICE_DEACTIVATED);
+  mobile_msg->set_connection_key(parent_->activated_connection_key());
+  mobile_msg->set_message_type(application_manager::MessageType::kNotification);
+  mobile_msg->set_correlation_id(message_.correlation_id());
+  mobile_msg->set_protocol_version(application_manager::ProtocolVersion::kV3);
```
Comment (KKarlash) [inline src/components/vr_cooperation/src/on_service_deactivated_notification.cc]: why kv3? why it set up here if it set in base class

File: src/components/vr_cooperation/include/vr_cooperation/vr_module_constants.h [hunk 5]
```diff
@@ -35,6 +35,12 @@
 
 namespace vr_cooperation {
 
+enum ServiceType {
+  kVoiceRecognition,
```
Comment (KKarlash) [inline src/components/vr_cooperation/include/vr_cooperation/vr_module_constants.h]: could be as const int with value 0

**Comment by VladSemenyuk** [comment]
Reviewed

**Comment by giangtqh** [comment]
@bonewell 
@KKarlash 
@thinhts 
Fix notes, add BaseCommandNotification, please review

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_command_notification.h [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+   */
+  virtual void OnTimeout() {}
+
+  /**
+   * @brief run request
+   */
+  virtual void Run();
+
+ protected:
+  vr_hmi_api::ServiceMessage message() {
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_command_notification.h]: I think in this case you can use reference on message like return type.

**Comment by bonewell** [comment]
Reviewed.

File: src/components/vr_cooperation/include/vr_cooperation/commands/on_service_deactivated_notification.h [hunk 7]
```diff
@@ -34,29 +34,30 @@
...
    * @brief Execute command
-   * @param message message to be sent to mobile
    */
-  void Execute(const application_manager::MessagePtr& message);
+  virtual void Execute();
 
  private:
-  VRModule* parent_;
-
   DISALLOW_COPY_AND_ASSIGN(OnServiceDeactivatedNotification);
```
Comment (thinhts) [inline src/components/vr_cooperation/include/vr_cooperation/commands/on_service_deactivated_notification.h]: why we need this?

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_command_notification.h (hunk 6 above)
Comment (thinhts) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_command_notification.h]: add discription

**Comment by thinhts** [comment]
Reviewed

File: src/components/vr_cooperation/include/vr_cooperation/commands/on_service_deactivated_notification.h (hunk 7 above)
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/commands/on_service_deactivated_notification.h]: Just for sure that no one will copy or assign object of this class :)

**Comment by giangtqh** [comment]
@thinhts 
@bonewell 
@KKarlash 
I fix your notes, please review

**Comment by bonewell** [comment]
Reviewed.

**Comment by KKarlash** [comment]
Reviewed

**Pull request merged**